tools/schemapi/vega_expr.py

```diff
@@ -0,0 +1,973 @@
+)
+
+if TYPE_CHECKING:
+    import sys
+    from re import Match, Pattern
+
+    from mistune import BlockState
+
+    if sys.version_info >= (3, 11):
+        from typing import LiteralString, Self
+    else:
+        from typing_extensions import LiteralString, Self
+    from _typeshed import SupportsKeysAndGetItem
+
+__all__ = ["parse_expressions", "write_expr_module"]
+
+
+# NOTE: Urls/fragments
+VEGA_DOCS_URL = "https://vega.github.io/vega/docs/"
+EXPRESSIONS_DOCS_URL = f"{VEGA_DOCS_URL}expressions/"
+
+
+class Source(str, enum.Enum):
+    LIVE = "https://raw.githubusercontent.com/vega/vega/main/docs/docs/expressions.md"
+    STATIC = "https://raw.githubusercontent.com/vega/vega/ff98519cce32b776a98d01dd982467d76fc9ee34/docs/docs/expressions.md"
+
+
+# NOTE: Regex patterns
+FUNCTION_DEF_LINE: Pattern[str] = re.compile(r"<a name=\"(.+)\" href=\"#(.+)\">")
+LIQUID_INCLUDE: Pattern[str] = re.compile(r"( \{% include.+%\})")
+SENTENCE_BREAK: Pattern[str] = re.compile(r"(?<!\.)\. ")
+
+# NOTE: `mistune` token keys/values
+TYPE: Literal[r"type"] = r"type"
+RAW: Literal["raw"] = "raw"
+SOFTBREAK: Literal["softbreak"] = "softbreak"
+TEXT: Literal["text"] = "text"
+CHILDREN: Literal["children"] = "children"
+
+# NOTE: Punctuation/markers
+ELLIPSIS: Literal["..."] = "..."
+OPEN_PAREN: Literal["("] = "("
+CLOSE_PAREN: Literal[")"] = ")"
+OPEN_BRACKET: Literal["["] = "["
+CLOSE_BRACKET: Literal["]"] = "]"
+INLINE_OVERLOAD: Literal[" |"] = " |"
+
+# NOTE: `altair` types (for annotations)
+RETURN_WRAPPER = "FunctionExpression"
+RETURN_ANNOTATION = "Expression"
+# NOTE: No benefit to annotating with the actual wrapper
+# - `Expression` is shorter, and has all the functionality/attributes
+CONST_WRAPPER = "ConstExpression"
+CLS_META = "_ExprMeta"
+INPUT_ANNOTATION = "IntoExpression"
+
+# NOTE: `python`/`mypy` related literals
+NONE: Literal[r"None"] = r"None"
+STAR_ARGS: Literal["*args"] = "*args"
+DECORATOR = r"@classmethod"
+IGNORE_OVERRIDE = r"# type: ignore[override]"
+IGNORE_MISC = r"# type: ignore[misc]"
+
+
+def _override_predicate(obj: Any, /) -> bool:
+    return (
+        callable(obj)
+        and (name := obj.__name__)
+        and isinstance(name, str)
+        and not (name.startswith("_"))
+    )
+
+
+_SCHEMA_BASE_MEMBERS: frozenset[str] = frozenset(
+    nm for nm, _ in getmembers(_SchemaBase, _override_predicate)
+)
+
+
+def download_expressions_md(url: str, /) -> Path:
+    """Download to a temporary file, return that as a ``pathlib.Path``."""
+    tmp, _ = request.urlretrieve(url)
+    fp = Path(tmp)
+    if not fp.exists():
+        msg = (
+            f"Expressions download failed: {fp!s}.\n\n"
+            f"Try manually accessing resource: {url!r}"
+        )
+        raise FileNotFoundError(msg)
+    else:
+        return fp
+
+
+def strip_include_tag(s: str, /) -> str:
+    """
+    Removes `liquid`_ templating markup.
+
+    .. _liquid:
+        https://shopify.github.io/liquid/
+    """
+    return LIQUID_INCLUDE.sub(r"", s)
```

Comment:
## TODO
- Investigate adding as an block/inline parser rule
- E.g. if this was a token `{"type": "liquid_template", ...}` it could be skipped much earlier

Reply:
Didn't have much luck with the plugin route.
The docs look pretty but didn't help much for this.

While trying to debug why my plugin didn't work, I settled on the solution in 
- [feat: Strip template markup earlier](https://github.com/vega/altair/pull/3600/commits/b2aeecb10008a8d6e57cf479be3b7d0f20097a5b)